packages/driver/src/util/serialization/log.ts

```diff
@@ -0,0 +1,400 @@
+    reifiedEl.setAttribute(attribute, props.attributes[attribute])
+  })
+
+  return reifiedEl
+}
+
+/**
+ * Attempts to take a generic data structure that is log-like and preprocess them for serialization. This generic may be/contain properties that are either
+ *  a) unserializable entirely
+ *  b) unserializable natively but can be processed to a serializable form (DOM elements or Functions)
+ *  c) serializable
+ *
+ * DOM elements are preprocessed via some key properties
+ * (attributes, classes, ids, tagName, value) including their innerHTML. Before the innerHTML is captured, inputs are traversed to set their stateful value
+ * inside the DOM element. This is crucial for body copy snapshots that are being sent to the primary domain to make the snapshot 'stateful'. Functions, if
+ * explicitly stated, will be preprocessed with whatever value they return (assuming that value is serializable). If a value cannot be preprocessed for whatever reason,
+ * null is returned.
+ *
+ *
+ * NOTE: this function recursively calls itself to preprocess a log
+ *
+ * @param {any} props a generic variable that represents a value that needs to be preprocessed before being sent through postMessage().
+ * @param {boolean} [attemptToSerializeFunctions=false] - Whether or not the function should attempt to preprocess a function by invoking it. USE WITH CAUTION!
+ * @returns {any} the serializable version of the generic.
+ */
+export const preprocessLogLikeForSerialization = (props, attemptToSerializeFunctions = false) => {
+  try {
+    if (_.isArray(props)) {
+      return props.map((prop) => {
+        try {
+          return preprocessLogLikeForSerialization(prop, attemptToSerializeFunctions)
+        } catch (e) {
+          return null
+        }
+      })
+    }
+
+    if (_.isPlainObject(props)) {
+      // only attempt to try and serialize dom elements and function (if attemptToSerializeFunctions is set to true)
+      let objWithPossiblySerializableProps = _.pickBy(props, (value) => {
+        const isSerializable = isSerializableInCurrentBrowser(value)
+
+        if (!isSerializable && $dom.isDom(value) || _.isFunction(value) || _.isObject(value)) {
+          return true
+        }
+
+        return false
+      })
+
+      let objWithOnlySerializableProps = _.pickBy(props, (value) => isSerializableInCurrentBrowser(value))
+
+      // assign the properties we know we can serialize here
+      let preprocessed: any = preprocessForSerialization(objWithOnlySerializableProps)
+
+      // and attempt to serialize possibly unserializable props here and fail gracefully if unsuccessful
+      _.forIn(objWithPossiblySerializableProps, (value, key) => {
+        try {
+          preprocessed[key] = preprocessLogLikeForSerialization(value, attemptToSerializeFunctions)
+        } catch (e) {
+          preprocessed[key] = null
+        }
+      })
+
+      return preprocessed
+    }
+
+    if ($dom.isDom(props)) {
```

Comment:
This is specifically for the $el attribute of the log correct?

Reply:
I'd expect to set the log's visibility here in case the dom ha changes when it's been reified.

Reply:
> I'd expect to set the log's visibility here in case the dom ha changes when it's been reified.

We really can't recheck the visibility of the element coming in because a) the test may have moved on by the time the event comes in to the primary and b) the primary may or may not have access to the DOM in this case. We have to take whatever the spec bridge tells us at the time of log added or changed to take for visibility, which will be accurate when we attach the logs associated snapshot(s) to the AUT after test end and the element is displayed in the console.